File: src/dojopool/static/js/multiplayer.ts
// Generated type definitions

function initMultiplayer(): any {
  // Implementation
}

function joinNearbyChat(): any {
  // Implementation
}

function sendChatMessage(): any {
  // Implementation
}

function addChatMessage(
  message: string,
  type: "user" | "system" = "user",
): void {
  // Implementation
}

function showChallengeModal(playerName: any, playerId: any): any {
  // Implementation
}

function sendChallenge(targetUserId: any): any {
  // Implementation
}

function showChallengeNotification(data: any): any {
  // Implementation
}

function respondToChallenge(challengeId: any, accept: any): any {
  // Implementation
}

function showChallengeStarted(data: any): any {
  // Implementation
}

function updateChallengeScore(data: any): any {
  // Implementation
}

function startChallengeTimer(): any {
  // Implementation
}

function showChallengeResults(data: any): any {
  // Implementation
}

// Type imports

let socket;
let currentRoom = null;
let activeChallenge = null;

// Initialize socket connection and event handlers
function initMultiplayer(): any {
  socket = io();

  socket.on("connect", () => {
    console.log("Connected to server");
    joinNearbyChat();
  });

  socket.on("user_joined", (data) => {
    addChatMessage(`${data.user} joined the area`, "system");
  });

  socket.on("new_message", (data) => {
    addChatMessage(`${data.user}: ${data.message}`);
  });

  socket.on("challenge_received", (data) => {
    if (data.challenger_id !== currentPlayerId) {
      showChallengeNotification(data);
    }
  });

  socket.on("challenge_started", (data) => {
    activeChallenge = {
      id: data.challenge_id,
      startTime: new Date(data.start_time),
      duration: data.duration,
      players: data.players,
    };
    showChallengeStarted(data);
    startChallengeTimer();
  });

  socket.on("challenge_declined", (data) => {
    if (activeChallenge && activeChallenge.id === data.challenge_id) {
      activeChallenge = null;
      showError(`${data.decliner} declined your challenge`);
    }
  });

  socket.on("score_updated", (data) => {
    if (activeChallenge && activeChallenge.id === data.challenge_id) {
      updateChallengeScore(data);
    }
  });

  socket.on("challenge_ended", (data) => {
    if (activeChallenge && activeChallenge.id === data.challenge_id) {
      showChallengeResults(data);
      activeChallenge = null;
    }
  });

  // Set up chat input handler
  const chatInput: any = document.getElementById("chat-input");
  const sendButton: any = document.getElementById("send-message");

  sendButton.addEventListener("click", () => {
    sendChatMessage();
  });

  chatInput.addEventListener("keypress", (e) => {
    if (e.key === "Enter") {
      sendChatMessage();
    }
  });

  // Set up challenge button handlers
  document.addEventListener("click", (e) => {
    if (e.target.classList.contains("challenge-btn")) {
      const playerItem: any = e.target.closest(".list-group-item");
      const playerName: any =
        playerItem.querySelector(".player-name").textContent;
      const playerId: any = playerItem.dataset.playerId;
      showChallengeModal(playerName, playerId);
    }
  });

  const sendChallengeBtn: any = document.getElementById("send-challenge");
  sendChallengeBtn.addEventListener("click", () => {
    const playerId: any = document.getElementById("challenge-player-id").value;
    sendChallenge(playerId);
  });
}

// Join nearby chat room based on current location
function joinNearbyChat(): any {
  if (!playerMarker) return;

  const pos: any = playerMarker.getPosition();
  fetch("/multiplayer/api/join-chat", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      lat: pos.lat(),
      lng: pos.lng(),
    }),
  })
    .then((response) => response.json())
    .then((data) => {
      if (data.status === "success") {
        currentRoom = data.room;
        addChatMessage("Joined local chat", "system");
      }
    })
    .catch((error) => {
      console.error("Error joining chat:", error);
      showError("Failed to join local chat");
    });
}

// Send chat message
function sendChatMessage(): any {
  const chatInput: any = document.getElementById("chat-input");
  const message: any = chatInput.value.trim();

  if (!message) return;

  fetch("/multiplayer/api/send-message", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      message: message,
    }),
  })
    .then((response) => response.json())
    .then((data) => {
      if (data.status === "success") {
        chatInput.value = "";
      }
    })
    .catch((error) => {
      console.error("Error sending message:", error);
      showError("Failed to send message");
    });
}

// Add message to chat container
function addChatMessage(
  message: string,
  type: "user" | "system" = "user",
): void {
  const chatContainer: any = document.getElementById("chat-messages");
  const messageDiv: any = document.createElement("div");
  messageDiv.className = `chat-message ${type}-message`;
  messageDiv.textContent = message;
  chatContainer.appendChild(messageDiv);
  chatContainer.scrollTop = chatContainer.scrollHeight;
}

// Show challenge modal
function showChallengeModal(playerName, playerId) {
  const modal: any = document.getElementById("challengeModal");
  document.getElementById("challenge-player-name").textContent = playerName;
  document.getElementById("challenge-player-id").value = playerId;
  new bootstrap.Modal(modal).show();
}

// Send challenge to nearby player
function sendChallenge(targetUserId: any): any {
  const duration: any = document.getElementById("challenge-duration").value;

  fetch("/multiplayer/api/challenge-player", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      target_user_id: targetUserId,
      duration: parseInt(duration),
    }),
  })
    .then((response) => response.json())
    .then((data) => {
      if (data.status === "success") {
        const modal: any = document.getElementById("challengeModal");
        bootstrap.Modal.getInstance(modal).hide();
        showSuccess("Challenge sent!");
      }
    })
    .catch((error) => {
      console.error("Error sending challenge:", error);
      showError("Failed to send challenge");
    });
}

// Show challenge notification
function showChallengeNotification(data: any): any {
  const notification: any = document.createElement("div");
  notification.className =
    "alert alert-info alert-dismissible fade show position-fixed top-50 start-50 translate-middle";
  notification.innerHTML = `
        ${data.challenger} challenged you to a ${data.duration / 60} minute coin collection competition!
        <div class="mt-2">
            <button class="btn btn-sm btn-success me-2" onclick="respondToChallenge('${data.challenge_id}', true)">Accept</button>
            <button class="btn btn-sm btn-danger" onclick="respondToChallenge('${data.challenge_id}', false)">Decline</button>
        </div>
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
  document.body.appendChild(notification);
}

// Respond to challenge
function respondToChallenge(challengeId, accept) {
  fetch("/multiplayer/api/respond-to-challenge", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      challenge_id: challengeId,
      accept: accept,
    }),
  })
    .then((response) => response.json())
    .then((data) => {
      if (data.status === "success" && accept) {
        showSuccess("Challenge accepted!");
      }
    })
    .catch((error) => {
      console.error("Error responding to challenge:", error);
      showError("Failed to respond to challenge");
    });
}

// Show challenge started notification
function showChallengeStarted(data: any): any {
  const opponent: any = data.players.find((p) => p.id !== currentPlayerId);
  showSuccess(
    `Challenge against ${opponent.name} started! Time remaining: ${data.duration / 60} minutes`,
  );

  // Create challenge scoreboard
  const scoreboard: any = document.createElement("div");
  scoreboard.id = "challenge-scoreboard";
  scoreboard.className =
    "position-fixed top-0 end-0 m-3 bg-dark p-3 rounded shadow";
  scoreboard.innerHTML = `
        <h5>Challenge Scoreboard</h5>
        <div id="challenge-timer" class="text-warning mb-2"></div>
        <div id="challenge-scores">
            ${data.players
              .map(
                (p) => `
                <div class="d-flex justify-content-between align-items-center mb-2">
                    <span>${p.name}</span>
                    <span id="score-${p.id}">0</span>
                </div>
            `,
              )
              .join("")}
        </div>
    `;
  document.body.appendChild(scoreboard);
}

// Update challenge score
function updateChallengeScore(data: any): any {
  const scoreElement: any = document.getElementById(`score-${data.player_id}`);
  if (scoreElement) {
    scoreElement.textContent = data.new_score;
  }
}

// Start challenge timer
function startChallengeTimer(): any {
  const timerElement: any = document.getElementById("challenge-timer");

  const updateTimer: any = () => {
    if (!activeChallenge) return;

    const now: any = new Date();
    const elapsed: any = (now - activeChallenge.startTime) / 1000;
    const remaining: any = activeChallenge.duration - elapsed;

    if (remaining <= 0) {
      timerElement.textContent = "Challenge ended!";
      return;
    }

    const minutes: any = Math.floor(remaining / 60);
    const seconds: any = Math.floor(remaining % 60);
    timerElement.textContent = `Time remaining: ${minutes}:${seconds.toString().padStart(2, "0")}`;

    requestAnimationFrame(updateTimer);
  };

  updateTimer();
}

// Show challenge results
function showChallengeResults(data: any): any {
  const scoreboard: any = document.getElementById("challenge-scoreboard");
  if (scoreboard) {
    scoreboard.remove();
  }

  const resultsModal: any = document.createElement("div");
  resultsModal.className = "modal fade";
  resultsModal.innerHTML = `
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                    <h5 class="modal-title">Challenge Results</h5>
                    <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                </div>
                <div class="modal-body">
                    <h4 class="text-center mb-4">${data.winner.name} wins!</h4>
                    <div class="final-scores">
                        ${Object.entries(data.scores)
                          .map(
                            ([id, score]) => `
                            <div class="d-flex justify-content-between align-items-center mb-2">
                                <span>${id === data.winner.id ? "👑 " : ""}${User.query.get(parseInt(id)).username}</span>
                                <span>$${score}</span>
                            </div>
                        `,
                          )
                          .join("")}
                    </div>
                </div>
            </div>
        </div>
    `;
  document.body.appendChild(resultsModal);
  new bootstrap.Modal(resultsModal).show();
}

// Initialize multiplayer features when the page loads
document.addEventListener("DOMContentLoaded", initMultiplayer);

File: src/blueprints/multiplayer.py
import logging
from flask import Blueprint, request, jsonify
from flask_socketio import emit, join_room, leave_room
from flask_login import current_user, login_required
from datetime import datetime, timedelta
from extensions import db
from models import User

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

multiplayer = Blueprint("multiplayer", __name__)

# Store active user sessions and their rooms
active_users = {}
user_rooms = {}
ongoing_challenges = {}

def get_nearby_room(lat, lng):
    """Generate a room name based on approximate location (100m grid)"""
    # Round coordinates to create a grid of roughly 100m squares
    grid_lat = round(lat * 1000) / 1000
    grid_lng = round(lng * 1000) / 1000
    return f"area_{grid_lat}_{grid_lng}"

@multiplayer.route("/api/join-chat", methods=["POST"])
@login_required
def join_chat():
    try:
        data = request.get_json()
        lat = float(data.get("lat"))
        lng = float(data.get("lng"))
        
        room = get_nearby_room(lat, lng)
        
        # Update user's current room
        if current_user.id in user_rooms:
            old_room = user_rooms[current_user.id]
            leave_room(old_room)
            
        user_rooms[current_user.id] = room
        join_room(room)
        
        # Notify others in the room
        emit("user_joined", {
            "user": current_user.username,
            "timestamp": datetime.utcnow().isoformat()
        }, room=room)
        
        return jsonify({"status": "success", "room": room})
    except Exception as e:
        logger.error(f"Error joining chat: {str(e)}")
        return jsonify({"status": "error", "message": str(e)}), 500

@multiplayer.route("/api/send-message", methods=["POST"])
@login_required
def send_message():
    try:
        data = request.get_json()
        message = data.get("message")
        room = user_rooms.get(current_user.id)
        
        if not room:
            return jsonify({"status": "error", "message": "Not in any chat room"}), 400
            
        emit("new_message", {
            "user": current_user.username,
            "message": message,
            "timestamp": datetime.utcnow().isoformat()
        }, room=room)
        
        return jsonify({"status": "success"})
    except Exception as e:
        logger.error(f"Error sending message: {str(e)}")
        return jsonify({"status": "error", "message": str(e)}), 500

@multiplayer.route("/api/challenge-player", methods=["POST"])
@login_required
def challenge_player():
    try:
        data = request.get_json()
        target_user_id = data.get("target_user_id")
        duration = int(data.get("duration", 300))  # Default 5 minutes
        
        room = user_rooms.get(current_user.id)
        if not room:
            return jsonify({"status": "error", "message": "Not in range of target player"}), 400
        
        # Create challenge data
        challenge_id = f"challenge_{current_user.id}_{target_user_id}_{datetime.utcnow().timestamp()}"
        ongoing_challenges[challenge_id] = {
            "challenger_id": current_user.id,
            "challenger_name": current_user.username,
            "target_id": target_user_id,
            "duration": duration,
            "status": "pending",
            "start_time": None,
            "scores": {str(current_user.id): 0}
        }
            
        emit("challenge_received", {
            "challenge_id": challenge_id,
            "challenger": current_user.username,
            "challenger_id": current_user.id,
            "duration": duration,
            "timestamp": datetime.utcnow().isoformat()
        }, room=room)
        
        return jsonify({"status": "success", "challenge_id": challenge_id})
    except Exception as e:
        logger.error(f"Error sending challenge: {str(e)}")
        return jsonify({"status": "error", "message": str(e)}), 500

@multiplayer.route("/api/respond-to-challenge", methods=["POST"])
@login_required
def respond_to_challenge():
    try:
        data = request.get_json()
        challenge_id = data.get("challenge_id")
        accept = data.get("accept", False)
        
        if challenge_id not in ongoing_challenges:
            return jsonify({"status": "error", "message": "Challenge not found"}), 404
            
        challenge = ongoing_challenges[challenge_id]
        if challenge["target_id"] != current_user.id:
            return jsonify({"status": "error", "message": "Not the challenge target"}), 403
            
        room = user_rooms.get(current_user.id)
        if not room:
            return jsonify({"status": "error", "message": "Not in a valid room"}), 400
            
        if accept:
            # Start the challenge
            challenge["status"] = "active"
            challenge["start_time"] = datetime.utcnow()
            challenge["scores"][str(current_user.id)] = 0
            
            emit("challenge_started", {
                "challenge_id": challenge_id,
                "players": [
                    {"id": challenge["challenger_id"], "name": challenge["challenger_name"]},
                    {"id": current_user.id, "name": current_user.username}
                ],
                "duration": challenge["duration"],
                "start_time": challenge["start_time"].isoformat()
            }, room=room)
        else:
            # Delete the challenge
            del ongoing_challenges[challenge_id]
            emit("challenge_declined", {
                "challenge_id": challenge_id,
                "decliner": current_user.username
            }, room=room)
            
        return jsonify({"status": "success"})
    except Exception as e:
        logger.error(f"Error responding to challenge: {str(e)}")
        return jsonify({"status": "error", "message": str(e)}), 500

@multiplayer.route("/api/update-challenge-score", methods=["POST"])
@login_required
def update_challenge_score():
    try:
        data = request.get_json()
        challenge_id = data.get("challenge_id")
        score_increment = int(data.get("score_increment", 0))
        
        if challenge_id not in ongoing_challenges:
            return jsonify({"status": "error", "message": "Challenge not found"}), 404
            
        challenge = ongoing_challenges[challenge_id]
        if str(current_user.id) not in challenge["scores"]:
            return jsonify({"status": "error", "message": "Not part of this challenge"}), 403
            
        # Check if challenge is still active
        if challenge["status"] != "active":
            return jsonify({"status": "error", "message": "Challenge is not active"}), 400
            
        elapsed_time = datetime.utcnow() - challenge["start_time"]
        if elapsed_time > timedelta(seconds=challenge["duration"]):
            # End the challenge
            challenge["status"] = "completed"
            room = user_rooms.get(current_user.id)
            
            # Determine winner
            scores = challenge["scores"]
            winner_id = max(scores.items(), key=lambda x: x[1])[0]
            winner = User.query.get(int(winner_id))
            
            emit("challenge_ended", {
                "challenge_id": challenge_id,
                "scores": scores,
                "winner": {
                    "id": winner.id,
                    "name": winner.username,
                    "score": scores[str(winner.id)]
                }
            }, room=room)
            
            return jsonify({"status": "success", "message": "Challenge ended"})
            
        # Update score
        challenge["scores"][str(current_user.id)] += score_increment
        
        # Broadcast score update
        room = user_rooms.get(current_user.id)
        emit("score_updated", {
            "challenge_id": challenge_id,
            "player_id": current_user.id,
            "new_score": challenge["scores"][str(current_user.id)]
        }, room=room)
        
        return jsonify({"status": "success"})
    except Exception as e:
        logger.error(f"Error updating challenge score: {str(e)}")
        return jsonify({"status": "error", "message": str(e)}), 500
